lib/fileset.nix

```diff
@@ -0,0 +1,993 @@
+            else
+              localTree == "directory";
+        in recurse rootComponentsLength sparseExtendedTree;
+
+    in
+    if ! isPath root then
+      if root._isLibCleanSourceWith or false then
+        throw ''
+          lib.fileset.toSource: Expected attribute `root` to be a path, but it's a value produced by `lib.sources` instead.
+              Such a value is only supported when converted to a file set using `lib.fileset.fromSource` and passed to the `fileset` attribute, where it may also be combined using other functions from `lib.fileset`.''
+      else if isCoercibleToString root then
+        throw ''
+          lib.fileset.toSource: Expected attribute `root` to be a path, but it's a string-like value instead, possibly a Nix store path.
+              Such a value is not supported, `lib.fileset` only supports local file filtering.''
+      else
+        throw "lib.fileset.toSource: Expected attribute `root` to be a path, but it's a ${typeOf root} instead."
+    else if ! pathExists root then
+      throw "lib.fileset.toSource: Expected attribute `root` \"${toString root}\" to be a path that exists, but it doesn't."
+    else if pathType root != "directory" then
+      throw "lib.fileset.toSource: Expected attribute `root` \"${toString root}\" to be a path pointing to a directory, but it's pointing to a file instead."
+    else if ! hasPrefix root fileset._base then
+      throw "lib.fileset.toSource: Expected attribute `fileset` to not be influenceable by any paths outside `root`, but `lib.fileset.getInfluenceBase fileset` \"${toString fileset._base}\" is outside `root`."
+    else
+      cleanSourceWith {
+        name = "source";
+        src = root;
+        filter = pathString: _: inSet (fromSubpath "./${pathString}");
+      };
+
+  /*
+  Create a file set from a filtered local source as produced by the `lib.sources` functions.
+  This does not import anything into the store.
+
+  Type:
+    fromSource :: SourceLike -> FileSet
+
+  Example:
+    fromSource (lib.sources.cleanSource ./.)
+  */
+  fromSource = source:
+    if ! source._isLibCleanSourceWith or false || ! source ? origSrc || ! source ? filter then
+      throw "lib.fileset.fromSource: Expected the argument to be a value produced from `lib.sources`, but got a ${typeOf source} instead."
+    else if ! isPath source.origSrc then
+      throw "lib.fileset.fromSource: Expected the argument to be source-like value of a local path."
+    else
+      _fromSource source.origSrc source.filter;
+
+  /*
+  Coerce a value to a file set:
+
+  - If the value is a file set already, return it directly
+
+  - If the value is a path pointing to a file, return a file set with that single file
+
+  - If the value is a path pointing to a directory, return a file set with all files contained in that directory
+
+  This function is mostly not needed because all functions in `lib.fileset` will implicitly apply it for arguments that are expected to be a file set.
```

Comment:
```suggestion
  This function will rarely be needed because all functions in `lib.fileset` will implicitly apply it for arguments that are expected to be a file set.
```